### PR TITLE
Support login tokens for multiple registries

### PR DESCRIPTION
**Pull request by cswindle**
This pull request builds on #4206 to support login using the the registry from alternate registries (RFC 2141). This includes the following changes:

 - allow credentials to be stored based on the registry
 - allow passing the registry to run cargo commands against using --registry

Note that this does not include a feature gate on the use of --registry as the publish code blocks publish if we use any features. @alexcrichton, are you happy with this approach, or is there a way that you would recommend this should be relaxed for testing purposes?

### CODE REVIEW

File: src/bin/login.rs [hunk 1]
```diff
@@ -44,26 +46,31 @@ pub fn execute(options: Options, config: &mut Config) -> CliResult {
...
+                }
+                None => {
+                    let src = SourceId::crates_io(config)?;
+                    let mut src = RegistrySource::remote(&src, config);
+                    src.update()?;
+                    let config = src.config()?.unwrap();
+                    options.flag_host.clone().unwrap_or(config.api.unwrap())
+                }
+            };
             println!("please visit {}me and paste the API Token below", host);
```
Comment (alexcrichton) [inline src/bin/login.rs]: Hm I think this should only get printed for the crates.io registry, right? We may not want to assume the layout of another registry per se just yet, and instead we could probably just return an error if no token was passed and `--registry` was passed?

File: src/cargo/ops/registry.rs [hunk 2]
```diff
@@ -189,24 +191,48 @@ fn transmit(config: &Config,
...
-    let index = config.get_string("registry.index")?.map(|p| p.val);
-    let token = config.get_string("registry.token")?.map(|p| p.val);
-    Ok(RegistryConfig { index: index, token: token })
+pub fn registry_configuration(config: &Config,
+                              registry: Option<String>) -> CargoResult<RegistryConfig> {
+
+    let (index, token) = match registry {
+        Some(registry) => {
+            let index = Some(config.get_registry_index(&registry)?);
+            let table = config.get_table(&format!("registry.{}", registry))?.map(|t| t.val);
```
Comment (alexcrichton) [inline src/cargo/ops/registry.rs]: Could this use `get_string` perhaps? That'll allow overrides via env var (table support isn't implemented there yet)

File: src/cargo/util/config.rs [hunk 3]
```diff
@@ -545,6 +545,14 @@ impl Config {
         }
     }
 
+    /// Gets the index for a registry.
+    pub fn get_registry_index(&self, registry: &str) -> CargoResult<String> {
```
Comment (alexcrichton) [inline src/cargo/util/config.rs]: Perhaps this could bake in returning a `Url` instead of a `String`?

**Comment by alexcrichton** [comment]
Looking great! Could this also be sure to place the support behind a feature gate? To do something like rustc it means we'd add a `-Z unstable-options` "feature" which would unlock other unstable flags like `--registry`

File: src/cargo/ops/registry.rs (hunk 2 above)
Comment (cswindle) [reply src/cargo/ops/registry.rs]: I did originally try using get_string, but that caused problems if there was a '.' in the registry name (which I think is allowed). I will have a play and see if I can get something working.

File: src/bin/login.rs (hunk 1 above)
Comment (cswindle) [reply src/bin/login.rs]: I was wondering if the /me API should be required for publish, I am happy to switch to just return an error stating that the token should be provided.

File: src/cargo/util/config.rs (hunk 3 above)
Comment (cswindle) [reply src/cargo/util/config.rs]: Sounds reasonable, I will update.

File: src/bin/login.rs (hunk 1 above)
Comment (alexcrichton) [reply src/bin/login.rs]: Yeah this is one where we haven't done much to sketch out the path towards "what does a custom registry need HTTP-wise", so I think it's ok to err on the side of caution for now and just return an error

File: src/cargo/ops/registry.rs (hunk 2 above)
Comment (cswindle) [reply src/cargo/ops/registry.rs]: Turns out it does just work switching to get_string, so I have made the change.

**Comment by cswindle** [comment]
@alexcrichton, I have now made all the changes that you requested as part of your review.

File: tests/alt-registry.rs [hunk 4]
```diff
@@ -262,3 +261,50 @@ fn alt_registry_and_crates_io_deps() {
...
+        .file("src/main.rs", "fn main() {}")
+        .build();
+
+    // Setup the registry by publishing a package
+    Package::new("bar", "0.0.1").alternative(true).publish();
+
+    // Now perform the actual publish
+    assert_that(p.cargo("publish").masquerade_as_nightly_cargo()
+                 .arg("--registry").arg("alternative").arg("-Zunstable-options"),
+                execs().with_status(101));
```
Comment (alexcrichton) [inline tests/alt-registry.rs]: Could you add an assertion about stderr here to ensure it prints the right error message?

File: tests/alt-registry.rs [hunk 5]
```diff
@@ -262,3 +261,50 @@ fn alt_registry_and_crates_io_deps() {
...
+
+    // Login so that we have the token available
+    assert_that(p.cargo("login").masquerade_as_nightly_cargo()
+                .arg("--registry").arg("alternative").arg("TOKEN").arg("-Zunstable-options"),
+                execs().with_status(0));
+
+    // Now perform the actual publish
+    assert_that(p.cargo("publish").masquerade_as_nightly_cargo()
+                 .arg("--registry").arg("alternative").arg("-Zunstable-options"),
+                execs().with_status(0));
```
Comment (alexcrichton) [inline tests/alt-registry.rs]: Hm could you describe the behavior of this test a bit? Presumably this is sending a `POST` which I guess is "uploading" a file to the local filesystem?

If that's the case, can this assert that the file was created? I think the binary format is pretty crazy so it's ok to not actually look inside the file, but asserting at least that it exists may be good?

File: tests/login.rs [hunk 6]
```diff
@@ -7,7 +7,7 @@ extern crate toml;
 use std::io::prelude::*;
```
Comment (alexcrichton) [inline tests/login.rs]: This (and I think at least one other file?) I think changed from 644 permissions to 755, mind reverting back?

File: tests/login.rs [hunk 7]
```diff
@@ -16,9 +16,16 @@ use cargo::core::Shell;
...
 const TOKEN: &str = "test-token";
+const ORIGINAL_TOKEN: &str = "api-token";
 const CONFIG_FILE: &str = r#"
     [registry]
     token = "api-token"
+
+    [registries.test-reg]
+    index = "http://dummy_index/"
+
+    [registries.test.reg]
```
Comment (alexcrichton) [inline tests/login.rs]: This actually reminds me, at some point we're going to want validation on the names of registries I think? Probably for now (to start conservative at least) we probably want to restrict registries to `[a-zA-Z0-9_-]` probably?

Comment (alexcrichton) [reply tests/login.rs]: Restrict the identifiers of the registry, that is

File: src/cargo/util/config.rs [hunk 8]
```diff
@@ -885,23 +895,36 @@ fn walk_tree<F>(pwd: &Path, mut walk: F) -> CargoResult<()>
...
+    };
+
+    let (key, value) = {
+        let key = "token".to_string();
+        let value = ConfigValue::String(token, file.path().to_path_buf());
+
+        if let Some(registry) = registry {
+            let mut map = HashMap::new();
+            map.insert(key, value);
+            (registry, CV::Table(map, file.path().to_path_buf()))
```
Comment (alexcrichton) [inline src/cargo/util/config.rs]: I think this is building up a TOML file like:

```toml
[my-registry-name]
token = "..."
```

right?


That seems fine by me for now (although maybe `registry.foo` for consistency with specifying the index?), but one thing I'm a little worried about is that we're making a lot of decisions in a bit of an ad-hoc fashion. I think all the conclusions are fine so far, but it'd be good to have a list of "here's the design decisions that were made" when we go for stabilizing custom registries.

Mind opening a tracking issue for custom registries (or appending to one that already exists?). I'm thinking we could track things like the `--registry` flag, my concern below for the naming of the registry identifiers, the format of the credentials file, the configuration syntax in `.cargo/config` for custom registries, etc.

Basically I'm thinking we should just track somewhere "here's what we need to be sure about for stabilization" and we can throw concerns/thoughts/implementation details on that as they arise.

File: tests/login.rs (hunk 7 above)
Comment (cswindle) [reply tests/login.rs]: In that case I will remove the test for test.reg and remove this entry. I will also add this to the issue list for stabilization.

File: tests/login.rs (hunk 6 above)
Comment (cswindle) [reply tests/login.rs]: I will revert those.

File: tests/alt-registry.rs (hunk 5 above)
Comment (cswindle) [reply tests/alt-registry.rs]: This is just performing an upload to the local registry in the same way that publish_clean currently works (with the only change that we are going to an alternate registry), I am not too keen on adding too much extra checking of the actual registry contents as that will likely be a remote server which we will be receiving the package normally. Are you happy for me to not make this change?

File: src/cargo/util/config.rs (hunk 8 above)
Comment (cswindle) [reply src/cargo/util/config.rs]: I will raise an issue to cover this.

File: tests/alt-registry.rs (hunk 5 above)
Comment (alexcrichton) [reply tests/alt-registry.rs]: Oh yeah I was just thinking we could assert that a file exists on the filesystem somewhere. Basically assert that the upload happened to the alternative registry here and not the "main registry"

**Comment by alexcrichton** [comment]
Ok looks great! r=me with an assertion that the publish happens to the right registry (asserting that a file exists)

**Comment by cswindle** [comment]
@alexcrichton, I have added the additional test that you asked for.

**Comment by cswindle** [comment]
Fixes #4498 and #3365.

**Comment by alexcrichton** [comment]
@bors: r+

Looks great!
